stellargraph/data/edge_cache.py

```diff
@@ -0,0 +1,206 @@
+    # No caching of edges
+    NONE = 1
+    # Cache edges, ignoring types
+    UNTYPED = 2
+    # TYPED = 3
+
+
+class EdgeCacheFactory(type):
+    """
+    Private class for instantiating the edge cache interface from
+    user-supplied information.
+    """
+
+    def __call__(cls, *args, **kwargs):
+        if cls is EdgeCache:
+            if len(args) > 0:
+                cache_type = args[0]
+                args = args[1:]
+            else:
+                cache_type = kwargs.pop("cache_type", None)
+            if cache_type is None:
+                # Supply default
+                cache_type = EdgeCacheType.UNTYPED
+            if not isinstance(cache_type, EdgeCacheType):
+                raise TypeError("Invalid cache type: {}".format(cache_type))
+            if cache_type == EdgeCacheType.NONE:
+                return NoEdgeCache(*args, **kwargs)
+            if cache_type == EdgeCacheType.UNTYPED:
+                return UntypedEdgeCache(*args, **kwargs)
+            raise ValueError("Unknown cache type: {}".format(cache_type))
+        else:
+            return type.__call__(cls, *args, **kwargs)
+
+
+class EdgeCache(metaclass=EdgeCacheFactory):
+    """
+    Encapsulation of node and edge data, for neighbourhood
+    queries.
+
+    Args:
+        cache_type (EdgeCacheType): The type of cache to use.
+        edge_data (EdgeData): The edge data object.
+        node_data (NodeData): The node data object.
+    """
+
+    def neighbour_nodes(self, node_id: Any) -> Iterable[Any]:
+        """
+        Obtains the collection of neighbouring nodes connected
+        to the given node, without regards to directionality.
+        Nodes may be repeated if there are multiple edges.
+
+        Args:
+            node_id (any): The identifier of the node in question.
+
+        Returns:
+            iterable: The identifiers of the neighbouring nodes.
+        """
+        raise NotImplementedError
+
+    def in_nodes(self, node_id: Any) -> Iterable[Any]:
+        """
+        Obtains the collection of neighbouring nodes with edges
+        directed to the given node. Nodes may be repeated if there
+        are multiple edges.
+
+        Args:
+            node_id (any): The identifier of the node in question.
+
+        Returns:
+            iterable: The identifiers of the neighbouring in-nodes.
+        """
+        raise NotImplementedError
+
+    def out_nodes(self, node_id: Any) -> Iterable[Any]:
+        """
+        Obtains the collection of neighbouring nodes with edges
+        directed from the given node. Nodes may be repeated if there
+        are multiple edges.
+
+        Args:
+            node_id (any): The identifier of the node in question.
+
+        Returns:
+            iterable: The identifiers of the neighbouring out-nodes.
+        """
+        raise NotImplementedError
+
+
+#############################################
+# Class for direct access to the edge data:
+
+
+class NoEdgeCache(EdgeCache):
```

Comment:
I'm not sure what circumstances we would not want to store the edge data as an adjacency list. What were your thoughts here?

Reply:
I was thinking that in the situation where memory is very tight, we might want to wear the increased time of access to the underlying data by turning off the additional caching.

Reply:
I understand your point – although this highlights a problem with the current design: For graphs with no edge features we store an edge list that has to be copied to an adjacency list for speed. To save memory we drop the adjacency list (the data structure we want for speed) leaving the edge list (the data structure we don't want). It would be better in this case to convert the edge list to an adjacency list and delete the edge list or directly ingest the adjacency list. 

Therefore I think it will be better to reduce memory and retain speed by looking at having an adjacency list but not the edge data. However, **I think we should leave these worries for the future and implement the minimum we require now**.